Add sea level curve function

xlia62 opened the pull request:
This PR adds support for specifying sea level curve in the FastScape Fortran plugin, allowing it to be defined either as a constant value or as a user-defined function. To demonstrate the impact of these options, two example figures are included showing topography transects: one with a constant sea level at –1 km (upper plot) and another with a sea level drop of 5 km (lower plot). Since there is currently no dedicated FastScape test in the /tests directory, a simple test input file (2d_fs_sea_level.prm.txt) is also provided to illustrate usage. 
![constant_sea_level](https://github.com/user-attachments/assets/d0eff30c-7947-45ef-ae99-59eabf6d44f9)
![dropping_sea_level](https://github.com/user-attachments/assets/d9709e3b-b807-4a0a-932a-b4b536091952)


* [ X] I have read the guidelines in our [CONTRIBUTING.md](../blob/main/CONTRIBUTING.md) document.
* [ X] I have followed the [instructions for indenting my code](../blob/main/CONTRIBUTING.md#making-aspect-better).
* [ X] I have tested my new feature locally to ensure it is correct.
* [X ] I have created a testcase in /tests for the new feature (2d_fs_sea_level.prm.txtl).
* [x] I have added a changelog entry in the  directory that will inform other users of my change.

naliboff reviewed (dismissed):
@xlia62 - Thanks for adding this feature, this will be incredibly useful! To first order what you have looks correct, but I think there are a number of simplifications that can be made to the code structure. Can you have a look through my comments, and then we can discuss how to proceed?

xlia62 commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -1898,7 +1925,25 @@ namespace aspect
           prm.enter_subsection("Marine parameters");
           {
-            sea_level = prm.get_double("Sea level");
+            const std::string format = prm.get("Sea level format");
```
Yes, modifications have been made. A boolean flag (Use sea level function) has been added. Users should still provide a Sea level value as before. However, if Use sea level function is set to true, they can now define a time-dependent sea level using a dealii parsed function.

xlia62 commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -503,6 +508,18 @@ namespace aspect
                                               &sediment_deposition_g,
                                               &slope_exponent_p);
+        // time to add some sediments                                       
+        if (use_marine_component)
```
Marine component is moved from initialization function. As sea level is updated in the loop so marine component will call the updated value.

naliboff reviewed (commented):
@xlia62 - Thanks for the changes. I've added a few more comments and suggestions to address.

xlia62 commented:
Thank you for reviewing again! I’ve accepted the suggested changes and added a new, simplified 2D test (2d_fs_test). This test uses a simple material setup and runs in just a few seconds. It’s now ready for another round of review.

gassmoeller commented on include/aspect/mesh_deformation/fastscape.h:
```diff
@@ -478,9 +478,29 @@ namespace aspect
+        /**
+         * sea level value that change with time.)
+         */
+        mutable double sea_level = 0.0; 
```
this should not be mutable either.

gassmoeller commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -429,6 +429,11 @@ namespace aspect
           // In later timesteps, we copy h directly from FastScape.
           std::mt19937 random_number_generator(fastscape_seed);
           std::uniform_real_distribution<double> random_distribution(-noise_elevation,noise_elevation);
+          // read sea level from the user defined function or constant value;
+          if (use_sea_level_function)
+            sea_level = sea_level_function.value(Point<1>(time_in_years));
+          else
+            sea_level = sea_level_constant_value;
```
If you change this to the following you dont need the `mutable` on the member variable:
```
const double current_sea_level = (use_sea_level_function)
?
sea_level_function.value(Point<1>(time_in_years))
:
sea_level_constant_value;
```

And you can do the same in the other place where you need the sea level.

gassmoeller reviewed (commented):
I have a few comments, but looking forward to this, I think it is a useful functionality.

naliboff commented on tests/2d_fs_test.prm.fastscape.prm:
```diff
@@ -0,0 +1,143 @@
+# This is a test on the input sea level function for the Fastscape Plugin. 
+# Sea level can be defined either as a constant value or as a user-defined function. 
+
+set Dimension                              = 2
```
@xlia62 - Can you provide a high-level summary of the test in terms of what features it is testing, overall, design, etc?

xlia62 commented on tests/2d_fs_test.prm.fastscape.prm:
```diff
@@ -0,0 +1,143 @@
+set Start time                             = 0
+set End time                               = 500e3
+set Use years in output instead of seconds = true
+set Nonlinear solver scheme                = no Advection, single Stokes 
```
Yes, it is.  And results have been added as well.

xlia62 commented on include/aspect/mesh_deformation/fastscape.h:
```diff
@@ -478,9 +478,29 @@ namespace aspect
+        /**
+         * Used defined constant sea level value (m).)
+         */
+        mutable double sea_level_constant_value = 0.0; 
```
Thank you for the suggestion. And now mutbale values have been changed and default values have been removed.

xlia62 commented:
Thank you for the review! Following the suggustion from @gassmoeller ,  the mutable variables have been updated to double, and the default values have been removed. The test case is fastscape_sea_level_function.prm, and the corresponding results can be found in tests/output-fastscape_sea_level_function.
Please feel free to review it at your convenience.

gassmoeller commented on tests/output-fastscape_sea_level_function/.DS_Store:
remove this file, it should not be part of ASPECT.

xlia62 commented:
Thank you for the note, the test prm and output have been modified. Please feel free to review at your convenience.

xlia62 commented:
Thank you for checking it  and the changelog is needed and unnecessary file is removed.. Ready for review again!

gassmoeller commented on doc/modules/changes/20250616_xlia62:
```diff
@@ -0,0 +1,5 @@
+Added: The ability for specifying sea level curve in the 
```
```suggestion
Added: The ability for specifying a sea level curve in the 
```

gassmoeller commented on doc/modules/changes/20250616_xlia62:
```diff
@@ -0,0 +1,5 @@
+Added: The ability for specifying sea level curve in the 
+FastScape Fortran plugin, allowing it to be defined either
+ as a constant value or as a user-defined function. 
```
```suggestion
as a constant or as a user-defined time-dependent function. 
```

gassmoeller reviewed (commented):
Almost ready to merge. Please:
- Address my two comments about the changelog entry, and
- rebase and squash all of your commits. Find me or @danieldouglas92 if you need help with that

bangerth commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -1898,7 +1934,19 @@ namespace aspect
           prm.enter_subsection("Marine parameters");
           {
-            sea_level = prm.get_double("Sea level");
+            use_sea_level_function = prm.get_bool("Use sea level function");
+            if (use_sea_level_function)
+            {
+              prm.enter_subsection("Sea level function");
+              {
+                sea_level_function.parse_parameters(prm);
+              }
+              prm.leave_subsection();
+            } 
+            else
+            {
+              sea_level_constant_value = prm.get_double("Sea level");
+            }
```
Like in the other PR, I think it would be useful to also "poison" the `sea_level_constant_value` if you're in the `if` branch:
```suggestion
            use_sea_level_function = prm.get_bool("Use sea level function");
            if (use_sea_level_function)
            {
              prm.enter_subsection("Sea level function");
              {
                sea_level_function.parse_parameters(prm);
              }
              prm.leave_subsection();
              
              // If using the function description for the sea level, no parts of the code
              // base should use the sea_level_constant_value variable. Poison it to 
              // make sure it really isn't used anywhere:
              sea_level_constant_value = numbers::signaling_nan<double>();
            } 
            else
            {
              sea_level_constant_value = prm.get_double("Sea level");
            }
```

bangerth commented on tests/fastscape_sea_level_function.prm:
```diff
@@ -0,0 +1,137 @@
+
+   subsection Marine parameters
+    set Use sea level function = true
+    subsection Sea level function
+      set Variable names = t
+      set Function expression = -1e-3*t
+    end 
```
This isn't right. What you are doing is saying that `t` is a *spatial* variable (in fact, the name of the only spatial variable given that you have a 1d function). But that is redefining what `t` stands for -- you can *always* use `t`, and it denotes time. I think this is right:
```suggestion
    subsection Sea level function
      set Function expression = -1e-3*t
    end 
```

bangerth commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -429,6 +429,11 @@ namespace aspect
           // In later timesteps, we copy h directly from FastScape.
           std::mt19937 random_number_generator(fastscape_seed);
           std::uniform_real_distribution<double> random_distribution(-noise_elevation,noise_elevation);
+          // read sea level from the user defined function or constant value;
+          const double current_sea_level = use_sea_level_function
+              ? sea_level_function.value(Point<1>(time_in_years))
+              : sea_level_constant_value;
```
This isn't right. What you want is this:
```suggestion
          // read sea level from the user defined function or constant value;
          sea_level_function.set_time(time_in_years);
          const double current_sea_level = use_sea_level_function
              ? sea_level_function.value(Point<1>()) // use a dummy point -- we only care about the time dependence
              : sea_level_constant_value;
```

xlia62 replied:
As this is a const method, I have to remove `sea_level_function.set_time(time_in_years);`. Instead, an update function is added to make this work in the latest commit

bangerth replied:
Oh, I see. That makes sense for the call to `set_time()`. But you should still call `.value()` with a dummy argument. The argument is a spatial point, and has nothing to do with time.

bangerth replied:
Scratch that -- I see that you did that already.

bangerth commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -780,8 +797,13 @@ namespace aspect
     {
       Assert (Utilities::MPI::this_mpi_process(this->get_mpi_communicator()) == 0, ExcInternalError());
-
+      // declare time in years
       const unsigned int current_timestep = this->get_timestep_number ();
+      const double time_in_years = this->get_time() / year_in_seconds;
+      // declare currrent sea level 
+      const double current_sea_level = use_sea_level_function
+          ? sea_level_function.value(Point<1>(time_in_years))
+          : sea_level_constant_value;
```
You need the same change as above.

bangerth reviewed (changes requested):
I've got a few additional comments where I think you could do things better.

Thanks for adding a FastScape test case -- we don't have enough of those!

xlia62 commented:
A new update method has been added to handle time updates. The commits have been rebased and squashed. Please feel free to take a look.

bangerth commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -795,8 +813,13 @@ namespace aspect
                                               std::vector<double> &silt_fraction) const
     {
       Assert (Utilities::MPI::this_mpi_process(this->get_mpi_communicator()) == 0, ExcInternalError());
-
+      // declare time in years
       const unsigned int current_timestep = this->get_timestep_number ();
```
I think the comment is now no longer relevant. Just remove it.

bangerth commented:
/rebuild

xlia62 commented:
Thank you again for looking at this. Those comments have been addressed. Feel free to move forward.
